
# 20250627 Adriaan van Os <adriaan@adriaan.biz>
# improve exception-handling on MacOS X
# See the description in rtl/inc/libexceptvars.pas below
# This file is for use with fpc-3.2.4rc1

diff -urN a/compiler/globtype.pas b/compiler/globtype.pas
--- a/compiler/globtype.pas	2020-09-15 21:40:36
+++ b/compiler/globtype.pas	2025-06-04 16:28:36
@@ -222,6 +222,7 @@
          cs_link_strip,cs_link_staticflag,cs_link_on_target,cs_link_extern,cs_link_opt_vtable,
          cs_link_opt_used_sections,cs_link_separate_dbg_file,
          cs_link_map,cs_link_pthread,cs_link_no_default_lib_order,
+         cs_external_exception_vars,
          cs_link_native,
          cs_link_pre_binutils_2_19,
          cs_link_vlink

diff -urN a/compiler/options.pas b/compiler/options.pas
--- a/compiler/options.pas	2021-05-01 16:09:57
+++ b/compiler/options.pas	2025-06-05 09:11:10
@@ -2608,7 +2611,8 @@
                       end;
                     'L' : begin  // -XLO is link order -XLA is link alias. -XLD avoids load defaults.
                                  // these are not aggregable.
-                            if (j=length(more)) or not (more[j+1] in ['O','A','D']) then
+                                 // -XLE (AvO) Exception vars are assumed to be external (Darwin)
+                            if (j=length(more)) or not (more[j+1] in ['O','A','D','E']) then
                               IllegalPara(opt)
                             else
                               begin
@@ -2623,7 +2627,8 @@
                                         if not LinkLibraryOrder.AddWeight(s) Then
                                            IllegalPara(opt);
                                        end;
-                                 'D' : include(init_settings.globalswitches,cs_link_no_default_lib_order)
+                                 'D' : include(init_settings.globalswitches,cs_link_no_default_lib_order);
+                                 'E' : include(init_settings.globalswitches,cs_external_exception_vars)
                                 else
                                   IllegalPara(opt);
                                  end; {case}
@@ -3725,6 +3719,10 @@
   { don't remove this, it's also for fpdoc necessary (FK) }
   def_system_macro('FPC_HAS_FEATURE_SUPPORT');
 
+{$ifdef DARWIN}
+  def_system_macro('FPC_EXTERNALEXCEPTVARS');
+{$endif}
+
   { make cpu makros available when reading the config files the second time }
   def_cpu_macros;

diff -urN a/compiler/pp.pas b/compiler/pp.pas
--- a/compiler/pp.pas	2019-11-09 23:07:33
+++ b/compiler/pp.pas	2025-06-04 16:28:36
@@ -211,7 +211,7 @@
     catch,
   {$endif}
 {$endif NOCATCH}
-  globals,compiler;
+  exceptvars, globals,compiler;
 
 var
   oldexit : pointer;

diff -urN a/compiler/scandir.pas b/compiler/scandir.pas
--- a/compiler/scandir.pas	2020-12-17 23:08:25
+++ b/compiler/scandir.pas	2025-06-04 16:28:36
@@ -139,6 +139,24 @@
       end;
 
 
+    procedure do_globalswitch(sw:tglobalswitch;optional:boolean);
+      var
+        state : char;
+      begin
+        if not current_module.in_global then
+         Message(scan_w_switch_is_global)
+        else
+          if optional then
+            state:=current_scanner.readoptionalstate('+')
+          else
+            state:=current_scanner.readstate;
+          if state='-' then
+            exclude(current_settings.globalswitches,sw)
+          else
+            include(current_settings.globalswitches,sw)
+      end;
+
+
     procedure do_message(w:integer);
       begin
         current_scanner.skipspace;
@@ -1892,6 +1910,10 @@
         do_localswitch(cs_zerobasedstrings);
       end;
 
+    procedure dir_externalexceptvars;
+      begin
+        do_globalswitch(cs_external_exception_vars,true)
+      end;
 
 {****************************************************************************
                          Initialize Directives
@@ -2030,6 +2052,7 @@
         AddDirective('Z2',directive_all, @dir_z2);
         AddDirective('Z4',directive_all, @dir_z4);
         AddDirective('ZEROBASEDSTRINGS',directive_all, @dir_zerobasesstrings);
+        AddDirective('EXTERNALEXCEPTVARS',directive_all, @dir_externalexceptvars);
       end;
 
 end.

diff -urN a/compiler/systems/t_darwin.pas b/compiler/systems/t_darwin.pas
--- a/compiler/systems/t_darwin.pas	2021-04-10 13:02:51
+++ b/compiler/systems/t_darwin.pas	2025-06-04 16:28:36
@@ -322,6 +322,7 @@
       HPath        : TCmdStrListItem;
       s,s1,s2      : TCmdStr;
       Fl1,Fl2      : Boolean;
+      HasExceptVars: Boolean;
     begin
       WriteResponseFile:=False;
       if ReOrderEntries Then
@@ -397,6 +398,7 @@
        end;
 
       { main objectfiles }
+      HasExceptVars:= False;
       LinkFilesFileName:=UniqueName('linkfiles')+'.res';
       FilesList:=TLinkRes.Create(outputexedir+LinkFilesFileName,false);
       while not ObjectFiles.Empty do
@@ -405,9 +407,19 @@
           if s<>'' then
             begin
               s:=TargetFixFileName(s);
+              if SameFileName( ExtractFileName( s), 'exceptvars.o')
+                then HasExceptVars := True;
+
               FilesList.Add(s);
             end;
         end;
+
+      if not HasExceptVars and not (cs_external_exception_vars in current_settings.globalswitches)
+        then begin
+          s:=TargetFixFileName('exceptvars.o');
+          FilesList.Add(FindObjectFile(s,'',false));
+        end;
+
       FilesList.writetodisk;
       FilesList.Free;

diff -urN a/compiler/utils/ppuutils/ppudump.pp b/compiler/utils/ppuutils/ppudump.pp
--- a/compiler/utils/ppuutils/ppudump.pp	2020-09-15 21:40:36
+++ b/compiler/utils/ppuutils/ppudump.pp	2025-06-04 16:28:36
@@ -2185,6 +2185,7 @@
         'Create linker map', {cs_link_map}
         'Link to pthread', {cs_link_pthread}
         'Link no default lib order', {cs_link_no_default_lib_order}
+        'Exception vars are implemented external (Darwin)', {cs_external_exception_vars}
         'Link using native linker', {cs_link_native}
         'Link for GNU linker version <=2.19', {cs_link_pre_binutils_2_19}
         'Link using vlink' {cs_link_vlink}

diff -urN a/packages/rtl-console/fpmake.pp b/packages/rtl-console/fpmake.pp
--- a/packages/rtl-console/fpmake.pp	2021-05-13 13:42:31
+++ b/packages/rtl-console/fpmake.pp	2025-06-04 16:28:36
@@ -2,7 +2,7 @@
 {$mode objfpc}{$H+}
 program fpmake;
 
-uses fpmkunit;
+uses exceptvars,fpmkunit;
 {$endif ALLPACKAGES}
 
 procedure add_rtl_console(const ADirectory: string);

diff -urN a/packages/rtl-console/src/unix/keyboard.pp b/packages/rtl-console/src/unix/keyboard.pp
--- a/packages/rtl-console/src/unix/keyboard.pp	2019-04-20 23:06:37
+++ b/packages/rtl-console/src/unix/keyboard.pp	2025-06-04 16:28:36
@@ -56,6 +56,7 @@
 {*****************************************************************************}
 
 uses
+  exceptvars,
   Mouse,  Strings,
   termio,baseUnix
   {$ifdef linux},linuxvcs{$endif};

diff -urN a/rtl/darwin/Makefile.fpc b/rtl/darwin/Makefile.fpc
--- a/rtl/darwin/Makefile.fpc	2018-07-10 12:25:58
+++ b/rtl/darwin/Makefile.fpc	2025-06-04 16:28:36
@@ -16,7 +16,8 @@
       charset cpall character getopts heaptrc lineinfo lnfodwrf lnfogdb \
       errors termio \
       sysconst cthreads rtlconsts cwstring bsd fpwidestring \
-      fpintres fpextres objc objcbase unixcp
+      fpintres fpextres objc objcbase unixcp \
+      exceptvars
 implicitunits=exeinfo \
       cp1250 cp1251 cp1252 cp1253 cp1254 cp1255 cp1256 cp1257 cp1258 \
       cp437 cp646 cp737 cp775 cp850 cp852 cp855 cp856 cp857 cp860 cp861 cp862 \
@@ -283,6 +284,9 @@
 	$(COMPILER) $<
 
 fpwidestring$(PPUEXT): $(OBJPASDIR)/fpwidestring.pp character$(PPUEXT) unixcp$(PPUEXT)
+	$(COMPILER) $<
+
+exceptvars$(PPUEXT) : $(INC)/exceptvars.pp $(SYSTEMUNIT)$(PPUEXT)
 	$(COMPILER) $<
 
 #

diff -urN a/rtl/inc/except.inc b/rtl/inc/except.inc
--- a/rtl/inc/except.inc	2015-03-28 14:09:36
+++ b/rtl/inc/except.inc	2025-06-18 14:45:20
@@ -22,9 +22,9 @@
 {$else FPC_HAS_FEATURE_THREADING}
 Var
 {$endif FPC_HAS_FEATURE_THREADING}
-  ExceptAddrStack   : PExceptAddr;
-  ExceptObjectStack : PExceptObject;
-  ExceptTryLevel    : ObjpasInt;
+  ExceptAddrStack   : PExceptAddr; external;
+  ExceptObjectStack : PExceptObject; external;
+  ExceptTryLevel    : ObjpasInt; external;
 
 Function RaiseList : PExceptObject;
 begin

diff -urN a/rtl/inc/exceptvars.pp b/rtl/inc/exceptvars.pp
--- a/rtl/inc/exceptvars.pp	1970-01-01 01:00:00
+++ b/rtl/inc/exceptvars.pp	2025-06-18 14:57:47
@@ -0,0 +1,86 @@
+
+{ adriaan van Os <adriaan@adriaan.biz> }
+
+{ 1. The problem
+
+  On MacOS, problems with exception handling occur if multiple binaries are compiled with FreePascal.
+  This is because each binary compiled with FreePascal has its own static copy of the runtime library.
+  This includes both code and data.
+  
+  The implies that, for example when a main program and a dynamic library are compiled, both will
+  have their own copy of the global variables involved in exception handling.
+  
+  Therefore, an exception frame in the main program and one in a library don't know of each other.
+  And that leads to incorrect exception handling.
+
+  2. The solution
+
+  The obvious solution to the above problem is to compile the FreePascal runtime library as a separate .dylib,
+  rather than to include it as static code. In this case, there is only one copy of both code and data of the
+  runtime-library. Unfortunately, building the FreePascal runtime-library as a dylib is broken and only some
+  rudiments from the past are still in the code and build machinery.
+
+  There are plans to add support to FreePascal to build the runtime-library as a package (which is a special
+  kind of dynamic library), but this is a work in progress.
+
+  3. A temporary solution
+
+  As a temporary solution, I have looked into
+
+  — isolating the global variables involved in exception handling from the FreePascal runtime-library and replacing
+    them by external variables
+
+  - introducing the -XLE command-line switch (and the equivalent $EXTERNALEXCEPTVARS compiler source directive) to
+    not (implicitely) link-in the standard solution.
+  
+  The standard solution (in this file in the runtime-library) doesn't address the exception var problem and simply
+  uses global variables. However, this file isn't linked in (inmplicitely) if the -XLE command-line
+  switch is passed or if the source code has the $EXTERNALEXCEPTVARS compiler directive. Thus, an external file can be
+  written (and added to the uses clause) that does solve the exception handling variables problems, replacing this file.
+}
+
+unit exceptvars;
+
+interface
+
+var
+  ErrorBase : Pointer = nil; public name 'FPC_ERRORBASE';
+  ErrorAddr : codepointer = nil; public;
+  ErrorCode : Word = 0; public;
+{ ErrorProc also ? }
+
+type
+  PCodePointer = ^codepointer;
+  ObjpasInt = Longint;
+
+{ contains the definition of the TExceptAddr type, which is required
+  by the compiler to generate code for any routine containing
+  implicit or explicit exceptions }
+{$i excepth.inc}
+
+{ Exception object stack, for Darwin }
+  type
+   PExceptObject = ^TExceptObject;
+   TExceptObject = record
+     FObject    : TObject;
+     Addr       : codepointer;
+     Next       : PExceptObject;
+     refcount   : Longint;
+     Framecount : Longint;
+     Frames     : PCodePointer;
+   end;
+
+{$ifdef FPC_HAS_FEATURE_THREADING}
+threadvar
+{$else FPC_HAS_FEATURE_THREADING}
+var
+{$endif FPC_HAS_FEATURE_THREADING}
+  ExceptAddrStack   : PExceptAddr; public;
+  ExceptObjectStack : PExceptObject; public;
+  ExceptTryLevel    : ObjpasInt; public;
+
+
+implementation
+
+begin
+end.

diff -urN a/rtl/inc/system.inc b/rtl/inc/system.inc
--- a/rtl/inc/system.inc	2020-02-04 00:16:03
+++ b/rtl/inc/system.inc	2025-06-18 15:02:11
@@ -48,8 +48,9 @@
 { Random / Randomize constants }
   OldRandSeed : Cardinal = 0;
 
+var
 { For Error Handling.}
-  ErrorBase : Pointer = nil;public name 'FPC_ERRORBASE';
+  ErrorBase : Pointer; external name 'FPC_ERRORBASE';
 
 { Used by the ansi/widestrings and maybe also other things in the future }
 var

diff -urN a/rtl/inc/systemh.inc b/rtl/inc/systemh.inc
--- a/rtl/inc/systemh.inc	2020-12-31 00:04:49
+++ b/rtl/inc/systemh.inc	2025-06-18 15:00:49
@@ -677,9 +677,12 @@
 
 { Exit Procedure handling consts and types  }
   ExitProc : codepointer = nil;
-  ErrorAddr: codepointer = nil;
-  ErrorCode: Word    = 0;
 
+var
+  ErrorAddr: codepointer; external;
+  ErrorCode: Word; external;
+
+const
 { file input modes }
   fmClosed = $D7B0;
   fmInput  = $D7B1;

diff -urN a/rtl/inc/libexceptvars.pas b/rtl/inc/libexceptvars.pas
--- a/rtl/inc/libexceptvars.pas	1970-01-01 01:00:00
+++ b/rtl/inc/libexceptvars.pas	2025-06-18 16:53:15
@@ -0,0 +1,94 @@
+
+{ Adriaan van Os <adriaan@adriaan.biz> }
+
+{ 1. The problem
+
+  On MacOS, problems with exception handling occur if multiple binaries are compiled with FreePascal.
+  This is because each binary compiled with FreePascal has its own static copy of the runtime library.
+  This includes both code and data.
+  
+  The implies that, for example when a main program and a dynamic library are compiled, both will
+  have their own copy of the global variables involved in exception handling.
+  
+  Therefore, an exception frame in the main program and one in a library don't know of each other.
+  And that leads to incorrect exception handling.
+
+  2. The solution
+
+  The obvious solution to the above problem is to compile the FreePascal runtime library as a separate .dylib,
+  rather than to include it as static code. In this case, there is only one copy of both code and data of the
+  runtime-library. Unfortunately, building the FreePascal runtime-library as a dylib is broken and only some
+  rudiments from the past are still in the code and build machinery.
+
+  There are plans to add support to FreePascal to build the runtime-library as a package (which is a special
+  kind of dynamic library), but this is a work in progress.
+
+  3. A temporary solution
+
+  As a temporary solution, I have looked into
+
+  — isolating the global variables involved in exception handling from the FreePascal runtime-library and replacing
+    them by external variables
+
+  - introducing the -XLE command-line switch (and the equivalent $EXTERNALEXCEPTVARS compiler source directive) to
+    not (implicitely) link-in the standard solution.
+  
+  The standard solution (in rtl/inc/exceptvars.pp) doesn`t address the exception var problem and simply
+  uses global variables. However, this file isn`t linked in (inmplicitely) if the -XLE command-line
+  switch is passed or if the source code has the $EXTERNALEXCEPTVARS compiler directive.
+  
+  The library solution in this file (rtl/inc/libexceptvars.pp) compiles the exception variables in a dynamic
+  library, making sure they exist only once, namely in the dynamic library. This requires using the -XLE command-line
+  switch or the $EXTERNALEXCEPTVARS compiler directive. It also requires linking in libexceptvars.dylib.
+}
+
+{$externalexceptvars}
+
+library libexceptvars;
+
+var
+  ErrorBase : Pointer = nil; public name 'FPC_ERRORBASE';
+  ErrorAddr : codepointer = nil; public;
+  ErrorCode : Word = 0; public;
+{ ErrorProc also ? }
+
+type
+  PCodePointer = ^codepointer;
+  ObjpasInt = Longint;
+
+{ contains the definition of the TExceptAddr type, which is required
+  by the compiler to generate code for any routine containing
+  implicit or explicit exceptions }
+{$i excepth.inc}
+
+{ Exception object stack, for Darwin }
+  type
+   PExceptObject = ^TExceptObject;
+   TExceptObject = record
+     FObject    : TObject;
+     Addr       : codepointer;
+     Next       : PExceptObject;
+     refcount   : Longint;
+     Framecount : Longint;
+     Frames     : PCodePointer;
+   end;
+
+{$ifdef FPC_HAS_FEATURE_THREADING}
+threadvar
+{$else FPC_HAS_FEATURE_THREADING}
+var
+{$endif FPC_HAS_FEATURE_THREADING}
+  ExceptAddrStack   : PExceptAddr; public;
+  ExceptObjectStack : PExceptObject; public;
+  ExceptTryLevel    : ObjpasInt; public;
+
+exports
+  ErrorBase,
+  ErrorAddr,
+  ErrorCode,
+  ExceptAddrStack,
+  ExceptObjectStack,
+  ExceptTryLevel;
+
+begin
+end.
